JENKINS-50556 don't check local refs if we have remote ones

jacob-keller opened the pull request:
JENKINS-50556 don't check local refs if we had remote ones

Fix an issue related to how polling works with remote vs local references. In the DefaultBuildChooser, we currently check branch-like references first, by converting them to remote/name lookups. This is done so that we can check the status of the branches according to the remote references. If we don't find one, then we assume that the branchSpec is actually a tag or tag-like reference.

Unfortunately, during polling, we prune these references that have already been built. This results in an accidental check of local branches, in the case where remote branches exist, but have no new changes. In some cases, this can result in polling reporting a change to build, but which will never be found by the build chooser during run time (as isPollCall would be false then, and we would not report the local branches in this case).

Fix this by instead pruning the list of references at the end, after collecting them. This ensures that the DefaultBuildChooser will work similarly in both the normal and polling cases, and avoids the possibility of permanently polling which can happen if a local branch gets a reference checked out.

Note that the polling bug may not be caught in many cases, because it happens that local branches do not get setup by the git plugin normally. Additionally, the remote polling setup is not susceptible to this issue, as it does not use the workspace for polling, instead using only ls-remote.

Add a test case which forces this sequence of events in order to highlight the issue, and make sure we don't re-introduce it.

jacob-keller commented:
I have a possible solution to add to this shortly.

jacob-keller commented:
Oops, the singleton collection returned by getHeadRevision is immutable, so we need to use addAll() when getting the tag-like references. Fixed that.

jacob-keller commented:
Ok, so there's some tests failing with timeout, and they *sometimes* fail locally, and other times they don't. Especially if I limit the test run to a single test case, it doesn't seem to fail. I also see similar tests fail locally if I run the test suite against master. Any thoughts on how to debug this?

jacob-keller commented:
It seems like some of the tests only appear to timeout during the full test run, but if run manually they do not timeout. Additionally, when I run the test suite locally, many tests timeout, but appear to have been marked as "flaky", and get re-run. I'm trying a local test run with the jenkins.test.timeout value set to a large number to see if it's maybe just a race condition for a few specific tests, or if it really is some deadlock that occurs.

jacob-keller commented:
I'm also pretty sure these tests are timing out due to the lack of entropy on the virtual machine (SSHD fails to start up when generating private keys due to blocking on /dev/random). It fails intermittently. Is there a way to re-run the tests?

MarkEWaite commented:
The simplest way that I've found to run tests on a pull request again is to `git commit --amend` the most recent commit, change the text of the commit message, then `git push --force` to update the pull request.

jacob-keller commented:
Looks like the same problem again with the failing to generate SSHD server keys in the JenkinsRule timeout.

jacob-keller commented:
Looks like @jglick tried something to re-run the tests,but they still timeout:

`Error
test timed out after 180 seconds
Stacktrace
org.junit.runners.model.TestTimedOutException: test timed out after 180 seconds
at java.lang.Object.wait(Native Method)
at java.lang.Object.wait(Object.java:502)
at org.jvnet.hudson.reactor.Reactor.execute(Reactor.java:267)
at jenkins.InitReactorRunner.run(InitReactorRunner.java:44)
at jenkins.model.Jenkins.executeReactor(Jenkins.java:916)
at jenkins.model.Jenkins.<init>(Jenkins.java:815)
at hudson.model.Hudson.<init>(Hudson.java:85)
at org.jvnet.hudson.test.JenkinsRule.newHudson(JenkinsRule.java:611)
at org.jvnet.hudson.test.JenkinsRule.before(JenkinsRule.java:384)
at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:537)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.lang.Thread.run(Thread.java:748)
Standard Output`

jacob-keller commented:
I fixed up the commit messages which needed to be re-worded, and squashed the test case into the same patch which fixes the bug. I also updated the message to clearly indicate the bug, and why we didn't find it in the past.

MarkEWaite commented:
I believe I've created a [verification job](https://github.com/MarkEWaite/docker-lfs/blob/lts-with-plugins/ref/jobs/Bugs-Individual/jobs/JENKINS-50556-workspace-poll-reports-changes/config.xml) that shows the problem in my environment.  Unfortunately, that job shows the same "build every poll" whether I'm using this set of changes or not.  

I still haven't found the correct set of configuration steps to duplicate the problem without the change and then to see the problem fixed with the change.

jacob-keller commented:
My test case is to have a job follow a branch "master" (but not explicitly marked as "origin/master"), and then also run a git checkout within the repository.

Specifically, I found this in a job which clones the repos, and makes a new commit on the branch and publishes it to gerrit.

I think your example has the */<name> which doesn't match a local branch reference.

If you look at my example test case, the only way so far I've found to trigger this is when the local job actually does a git checkout of some kind, as currently we happen to avoid the issue due to how clone never actually checks out the branch but instead just leaves us on a detached head checkout, and local branches aren't fetched.

So, the only way I've actually triggered this is to explicitly checkout something in a shell script or by hand in the workspace outside the job. I suppose there may also be configurations which do result in a checkout of the local branch, but I haven't found one yet.

Additionally, the checkout would need to be to some sha1 which hasn't yet been built by some other job. Otherwise we'll check against build data and discover it was already built by a different build.

MarkEWaite commented:
I'll experiment further.  I'm using the "LocalBranch" extension (checkout to specific local branch) which performs a checkout) and forcing workspace polling.  I'll try switching the definition of the branch to checkout from the intentionally wildcard version to a non-wildcard version.  

Thanks for investigating!
